File: storagemanager/src/main/java/com/example/alexeyglushkov/cachemanager/clients/Cache.java
package com.example.alexeyglushkov.cachemanager.clients;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * Created by alexeyglushkov on 03.03.18.
 */

public interface Cache {
    enum CacheMode {
        CHECK_CACHE_IF_ERROR_THEN_LOAD,
        IGNORE_CACHE,
        ONLY_LOAD_FROM_CACHE,
        ONLY_STORE_TO_CACHE
    }

    void putValue(@NonNull String key, @NonNull Object value) throws Exception;
    @Nullable <T> T getCachedValue(@NonNull String cacheKey) throws Exception;

    CacheMode getCacheMode();
    void setCacheMode(CacheMode mode);

    //// Inner classes
    class CacheEmptyError extends Error {
        private static final long serialVersionUID = -783104001989492147L;

        private String cacheKey;

        public CacheEmptyError(String cacheKey, Throwable throwable) {
            super("Cache is empty", throwable);
            this.cacheKey = cacheKey;
        }

        public String getCacheKey() {
            return cacheKey;
        }
    }
}

File: storagemanager/src/main/java/com/example/alexeyglushkov/cachemanager/clients/Cache.kt
package com.example.alexeyglushkov.cachemanager.clients

/**
 * Created by alexeyglushkov on 03.03.18.
 */
// TODO: think about task manager based cache
interface Cache {
    enum class CacheMode {
        CHECK_CACHE_IF_ERROR_THEN_LOAD, IGNORE_CACHE, ONLY_LOAD_FROM_CACHE, ONLY_STORE_TO_CACHE
    }

    @Throws(Exception::class) fun putValue(key: String, value: Any)
    @Throws(Exception::class) fun <T> getCachedValue(cacheKey: String): T?

    var cacheMode: CacheMode

    //// Inner classes
    class CacheEmptyException(val cacheKey: String, throwable: Throwable?) : Exception("Cache is empty", throwable) {
        companion object {
            private const val serialVersionUID = -783104001989492147L
        }
    }
}